project_generator/project.py

```diff
@@ -169,68 +170,61 @@ def __init__(self, name, project_files, pgen_workspace):
         """initialise a project with a yaml file"""
         self.project = {}
         self.workspace = pgen_workspace
-
+        self.tool_specific = defaultdict(ToolSpecificSettings)
         self.name = name
-
-        self.include_paths = []
-        self.source_paths = []
-        self.source_groups = {}
-        self.macros = []
-        self.project_dir = {
-            'name': '',
-            'path': '',
-        }
-        self.build_dir = 'build'
-        self.output_dir = {
-            'path': '',
-            'rel_path': '',
-            'rel_count': '',
-        }
-        self.mcu = ''
-        self.core = ''
-        self.target = ''
-        self.tools_supported = []
-
         self.output_types = {
             'executable': 'exe',
             'exe': 'exe',
             'library': 'lib',
             'lib': 'lib',
         }
-        self.output_type = self.output_types['executable']
-
-        self.debugger = 'cmsis-dap'
-
-        self.linker_file = None
-        self.tool_specific = defaultdict(ToolSpecificSettings)
-
-        self.project_name = None
         self.tools = ToolsSupported()
+        self.source_groups = {}
+        self._fill_defaults()
 
         for project_file in project_files:
             try:
                 f = open(project_file, 'rt')
                 project_file_data = yaml.load(f)
-                self.set_attributes(project_file_data)
+                self._set_attributes(project_file_data)
             except IOError:
                raise IOError("The file %s referenced in main yaml doesn't exist."%project_file)
 
-        if self.project_dir['path'] == '':
-            self.project_dir['path'] = self.workspace.settings.generated_projects_dir_default
+    def _fill_defaults(self):
+        self.project['name'] = self.name
+        self.project['output_type'] = self.output_types['executable']
+        self.project['output_type'] = self.output_types['executable']
+        self.project['linker_file'] = None
+        self.project['project_name'] = None
+        self.project['tools_supported'] = [self.workspace.settings.DEFAULT_TOOL]
+        self.project['debugger'] = 'cmsis-dap'
+        self.project['build_dir'] = 'build'
+        self.project['output_dir'] = {
+            'path': '',
+            'rel_path': '',
+            'rel_count': '',
+        }
+        self.project_dir = {
+            'name': '',
+            'path': self.workspace.settings.generated_projects_dir_default,
+        }
+        self.project['misc'] ={}
+        self.path_arrays_keys = ['includes', 'macros', 'source_paths']
```

Comment:
Yeah, I moved that down into the for loop because of the confusing name, just a quick way to declare all of those as arrays.